narciuch/assignment2/src/TreeAncestors.java

```diff
@@ -0,0 +1,135 @@
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+
+public class TreeAncestors<T> {
+
+  public void printAncestors(T key, Node treeRoot) {
+    List<Node> result = getAncestors(key, treeRoot);
+    printNodeList(result);
+  }
+
+  /**
+   * Returns a list of nodes that are parents, grandparents etc, of the node
+   * with the value of 'key'.
+   *
+   * @param key : The value of the node whos ancestors you want to find
+   * @param node : The root of the tree in which you want to search for the ancestors
+   * @return : List of nodes which are the ancestors of the node with value 'key'.
+   *           These are in the order of parent, grandparent, ..., root
+   */
+  public List<Node> getAncestors(T key, Node node) {
+    if (node == null) {
+      return Collections.emptyList();
+    }
+    List<Node> path = new ArrayList<>();
+    boolean result = getAncestorsHelper(key, node, path);
+    Collections.reverse(path); //Puts the ancestors into the desired order
+    return path;
+  }
+
+  /**
+   * Used to find the path from the root of a tree to a particular node
+   * (i.e. the nodes ancestors)
+   *
+   * @param key : The value of the desired node
+   * @param node : Tree to search through
+   * @param path : Indicates the visited nodes in the path from the original root directly to
+   *             the current 'node'
+   * @return true if the node with value 'key' was found in the tree 'node'
+   */
+  public boolean getAncestorsHelper(T key, Node node, List<Node> path) {
+    if (node == null) {
+      return false;
+    }
+    if (node.getValue() == key) {
+      return true;
+    }
+    boolean left = false;
+    boolean right = false;
+    if (node.getLeft() != null) {
+      path.add(node);
+      left = getAncestorsHelper(key, node.getLeft(), path);
+    }
+    if (node.getRight() != null) {
+      path.add(node);
+      right = getAncestorsHelper(key, node.getRight(), path);
+    }
+    if (path.size() != 0 && !left && !right) {
```

Comment:
Current code does not seem to enforce uniqueness of keys and there is no comment that would suggest that it is a requirement, so it is possible that node with given value is present in both left and right subtree.
What will happen in such case?

Comment is about the method in general, not necessarily line 58 :)

Reply:
Very good point.
Have added a test and comment to getAncestors() about the uniqueness of keys.

As for the comment on line 58, it seemed to be more of an implementation detail, explaining exactly what is happening within that part of the code, irrelevant to the rest of the function. 
I'm not quite sure what you mean here.